event/event.go

```diff
@@ -50,3 +51,62 @@ func (p *Publisher[T]) Publish(ctx context.Context, event T) error {
 		Body: encBody,
 	})
 }
+
+// RawSubscription represents a subscription that delivers messages as is.
+// No assumptions are made about the message contents. This should rarely be used in favor of [Subscription].
+type RawSubscription struct {
+	sub            *pubsub.Subscription
+	maxConcurrency int
+}
+
+// RawMessageHandler is responsible for handling raw messages from a subscription.
+type RawMessageHandler func([]byte) error
+
+// NewRawSubscription creates a new raw subscription. It provides messages in a
+// service like manner (serve) and manages concurrent execution, each message
+// is processed in its own goroutines respecting the given maxConcurrency.
+func NewRawSubscription(url string, maxConcurrency int) (*RawSubscription, error) {
+	if maxConcurrency <= 0 {
+		return nil, fmt.Errorf("max concurrency must be > 0: %d", maxConcurrency)
+	}
+	// We dont want the subscription to expire, so we use the background context.
+	sub, err := pubsub.OpenSubscription(context.Background(), url)
+	if err != nil {
+		return nil, err
+	}
+	return &RawSubscription{
+		sub:            sub,
+		maxConcurrency: maxConcurrency,
+	}, nil
+}
+
+// Serve will start serving all messages from the subscription calling handler for each
+// message. It will run until [RawSubscription.Shutdown] is called.
+// If the error is nil Ack is sent.
+// If a non-nil error is returned by the handler Unack will be sent.
+// Serve may be called multiple times, each time will start a new serving service that will
+// run up to "maxConcurrency" goroutines.
+func (r *RawSubscription) Serve(handler RawMessageHandler) error {
+	semaphore := make(chan struct{}, r.maxConcurrency)
+	for {
+		semaphore <- struct{}{}
```

Comment:
Interesting enough the semaphore example from the lib seems broken: https://pkg.go.dev/gocloud.dev@v0.30.0/pubsub#example-Subscription.Receive-Concurrent

It first receives a message and then acquires the semaphore. If you have to wait for minutes on the semaphore the message will timeout since you acquired the message first. Makes more sense to first of all acquire the semaphore, then do anything.

Unless Im missing something.

Reply:
Yep, it seems so. they are doing the semaphore to handle the message not to read the message.